driver/normalizer/annotation.go

```diff
@@ -1,34 +1,368 @@
+	AnnotateType("CPPASTTemplateId", nil, role.Type, role.Incomplete),
+	AnnotateType("CPPASTDeleteExpression", nil, role.Call, role.Expression, role.Incomplete),
+	AnnotateType("CPPASTInitializerList", nil, role.Initialization, role.List),
+	AnnotateType("CPPASTCastExpression", nil, role.Expression, role.Incomplete),
+	AnnotateType("CPPASTDesignatedInitializer", nil, role.Expression, role.Initialization),
+	AnnotateType("CPPASTConditionalExpression", nil, role.Expression, role.Condition),
+
+	AnnotateTypeCustom("CPPASTUnaryExpression",
+		FieldRoles{
+			"operator": {Op: Var("operator")},
+		},
+		LookupArrOpVar("operator", unaryExprRoles)),
+
+	AnnotateTypeCustom("CPPASTSimpleDeclSpecifier",
+		FieldRoles {
+			"Type": {Rename: uast.KeyToken, Op: Var("type")},
+		},
+		LookupArrOpVar("type", typeRoles),
+	),
+
+	AnnotateType("CPPASTASMDeclaration", FieldRoles{
+		"Assembly": {Rename: uast.KeyToken},
+	}, role.Declaration, role.Block, role.Incomplete),
+
+	AnnotateType("CPPASTLinkageSpecification", FieldRoles{
+		"Literal": {Rename: uast.KeyToken},
+	}, role.Declaration, role.Block, role.Incomplete),
+
+	AnnotateTypeCustom("CPPASTLiteralExpression",
+		FieldRoles {
+			"LiteralValue": {Rename: uast.KeyToken},
+			"kind": {Op: Var("kind")},
+		},
+		LookupArrOpVar("kind", litExprRoles),
+	),
+
+	AnnotateType("CPPASTCompoundStatement", nil, role.Body),
+	AnnotateType("CPPASTDeclarator", FieldRoles{
+		"Name": {Rename: uast.KeyToken},
+	}, role.Declaration, role.Variable, role.Name),
+	AnnotateType("CPPASTDeclarator", nil, role.Declaration, role.Variable, role.Name),
+
+	AnnotateType("CPPASTFunctionDefinition", ObjRoles{
+		"Prop_Body": {role.Function, role.Declaration, role.Body},
+		"Prop_DeclSpecifier": {role.Function, role.Declaration, role.Return, role.Type},
+	}, role.Function, role.Declaration),
+
+	AnnotateType("CPPASTFunctionDeclarator", FieldRoles{
+		"Prop_Name": {Roles: role.Roles{role.Function, role.Declaration, role.Name}},
+		// SDK TODO: adding "Opt: true" fails since Arrays can't be optional, but without it the annotation won't
+		// match, thus the duplicated annotation below
+		"Prop_Parameters": {Arr: true, Roles: role.Roles{role.Function, role.Declaration, role.Argument}},
+	}, role.Function, role.Declaration),
+
+	AnnotateType("CPPASTFunctionDeclarator", FieldRoles{
+		"Prop_Name": {Roles: role.Roles{role.Function, role.Declaration, role.Name}},
+	}, role.Function, role.Declaration),
+
+	AnnotateType("CPPASTReturnStatement", ObjRoles{
+		"Prop_ReturnArgument": {role.Return, role.Value},
+	}, role.Statement, role.Return),
+
+	AnnotateTypeCustom("CPPASTBinaryExpression", MapObj(Obj{
+		"Operator": Var("operator"),
+		"Prop_Operand1": ObjectRoles("operand1"),
+		"Prop_Operand2": ObjectRoles("operand2"),
+		// Temporarily using the same name to detect if those are really duplicated
+		"Prop_InitOperand2": ObjectRoles("operand2"),
```

Comment:
Hmm, I'm not sure if it will do the correct thing here. The `ObjectRoles` is pretty complex, I don't think it will deduplicate correctly. As a side effect, it may append roles twice. But it might be okay since we will deduplicate roles anyway.

Reply:
Mmm, that's my interpretation of what you suggested. Probably can be make easier, but I don't see it being so complex, do you have a simpler alternative in mind? I tested that it deduplicated the roles and it worked (especially considering that this node is everywhere).

Reply:
By "complex" I mean an implementation of `ObjectRoles` :)
Not, I don't have anything better in mind, so let's leave it as-is if it works. I'm sure it will trigger some error if this field is not a duplicate.